src/drivers/uavcan_v1/templates/_composite_type.j2

```diff
@@ -0,0 +1,115 @@
+{{ composite_type | definition_begin }}
+{
+{%- if T.fixed_port_id is not None %}
+	static constexpr CanardPortID PORT_ID = {{T.fixed_port_id}};
+{%- endif %}
+
+{%- set total_size = namespace(value=0) -%}
+{%- for field in composite_type.fields -%}
+    {% set total_size.value = total_size.value + field.data_type.bit_length_set|max %}
+{%- endfor %}
+	static constexpr size_t SIZE = {{total_size.value}};
+
+	static constexpr auto getDataTypeFullName() { return "{{T}}"; }
+
+{% for constant in composite_type.constants %}
+	static constexpr {{ constant.data_type | declaration }} {{ constant.name | id }} = {{ constant.value.native_value.numerator }} / {{ constant.value.native_value.denominator }};
+{%- endfor -%}
+
+{% if composite_type is UnionType %}
+#error "TODO: UnionType
+{% else %}
+
+{% for field in composite_type.fields -%}
+    {%- if field is not padding %}
+        {%- if field.data_type is FloatType %}
+            {%- if field.data_type.bit_length_set|max <= 32 %}
+	float {{ field | id }}{NAN};
+            {%- else %}
+	double {{ field | id }}{NAN};
+            {%- endif %}
+        {%- elif field.data_type is BooleanType %}
+	bool {{ field | id }}{false};
+        {%- else %}
+	{{ field.data_type | declaration }} {{ field | id }}{};
+        {%- endif -%}
+    {%- endif -%}
+{%- endfor %}
+{% endif %}
+
+	void serializeToBuffer(uint8_t* const buffer, const size_t starting_bit = 0)
+	{
+{%- set bit_offset = namespace(value=0) -%}
+{%- for field in composite_type.fields -%}
+    {%- if field is not padding %}
+      {%- if field.data_type is SerializableType %}
+        {%- if field.data_type is IntegerType %}
+		canardDSDLSetUxx(buffer, starting_bit + {{ bit_offset.value }}, {{ field.name }}, {{ field.data_type.bit_length_set | max }});
+        {%- elif field.data_type is BooleanType %}
+		canardDSDLSetBit(buffer, starting_bit + {{ field.data_type.bit_length_set | max }}, {{ field.name }});
+        {%- elif field.data_type is FloatType %}
+            {%- if field.data_type.bit_length_set == 16 %}
+		canardDSDLSetF16(buffer, starting_bit + {{ field.data_type.bit_length_set | max }}, {{ field.name }});
+            {%- elif field.data_type.bit_length_set == 32 %}
+		canardDSDLSetF32(buffer, starting_bit + {{ field.data_type.bit_length_set | max }}, {{ field.name }});
+            {%- elif field.data_type.bit_length_set == 64 %}
+		canardDSDLSetF64(buffer, starting_bit + {{ field.data_type.bit_length_set | max }}, {{ field.name }});
+            {%- endif %}
+        {%- else %}
+		{{ field.name }}.serializeToBuffer(buffer, starting_bit + {{ bit_offset.value }});
```

Comment:
bit_offset is a set, not a scalar. This logic will break if the type contains at least one variable-length entity such as an array or a union. Bit offset is a runtime variable and it shall be modeled as such by the generated code. This comment applies to the deserialization code as well.